### PR TITLE
Resolve sampling memory inefficiencies for HigherOrderGP

### PR DESCRIPTION
**Pull request by wjmaddox**
Resolves some of the sampling memory inefficiences for the HigherOrderGP. I'd originally implemented the posterior sampling with a trailing batch dimension so that the base samples were being stored as `num_samples x batch x dim x 1`. Surprisingly, this is somewhat less memory efficient than the gpytorch way of storing base samples `batch x dim x num_samples`. This PR resolves that issue and cleans up some of the funkiness associated with that trailing dimension. I think it'll make backpropagation through the rsample somewhat faster as well.

Example that now works:
```python
import torch
from torch import Tensor
from botorch import fit_gpytorch_model
from botorch.acquisition import MCAcquisitionObjective, qExpectedImprovement
from botorch.models.higher_order_gp import HigherOrderGP, FlattenedStandardize
from botorch.models import SingleTaskGP
from botorch.optim import optimize_acqf
from botorch.sampling import IIDNormalSampler
import gpytorch
from gpytorch import ExactMarginalLogLikelihood

device = "cuda" if torch.cuda.is_available() else "cpu"

train_X = torch.rand(100, 2, device=device)
train_Y = torch.rand(100, num_outputs, device=device)
model = HigherOrderGP(
    train_X,
    train_Y,
    # outcome_transform=FlattenedStandardize(output_shape=torch.Size([num_outputs]))
)
mll = ExactMarginalLogLikelihood(model.likelihood, model)
_ = fit_gpytorch_model(mll)

class SumSquaresObjective(MCAcquisitionObjective):
    def forward(self, samples: Tensor) -> Tensor:
        return samples.pow(2).sum(dim=-1)


f_star = SumSquaresObjective()(train_Y).max()

acqf = qExpectedImprovement(
    model=model,
    best_f=f_star,
    objective=SumSquaresObjective(),
    sampler=IIDNormalSampler(512),
)

sol, _ = optimize_acqf(
    acq_function=acqf,
    bounds=torch.tensor([[0.], [1.]]).repeat(1, 2).to(device=device),
    q=1,
    num_restarts=10,
    raw_samples=40,
)
```

## Motivation

Provides a partial resolution to #644. Thanks for pointing this issue out! cc @saitcakmak @Balandat 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Unit tests pass.

## Related PRs

#646

### CODE REVIEW

File: test/models/test_higher_order_gp.py [hunk 1]
```diff
@@ -257,19 +257,10 @@ def test_posterior(self):
...
+            sampled_variance = samples_3.var(dim=0).view(-1)
             posterior_variance = posterior_variance.view(-1)
             self.assertLess(
                 (posterior_variance - sampled_variance).norm()
                 / posterior_variance.norm(),
                 5e-2,
             )
-
-
-# def TestFlattenedStandardize(BotorchTestCase):
```
Comment (Balandat) [inline test/models/test_higher_order_gp.py:269]: we should probably add an atomic test for this as well, but that doesn't need to be part of this PR

File: botorch/models/higher_order_gp.py [hunk 2]
```diff
@@ -291,16 +279,17 @@ def rsample(
...
+        covar_root = covariance_matrix.root_decomposition().root
         samples = covar_root.matmul(base_samples)
 
         # now pluck out Y_x and X_x
         noiseless_train_marginal_samples = samples[
             ..., : self.train_train_covar.shape[-1], :
         ]
         test_marginal_samples = samples[..., self.train_train_covar.shape[-1] :, :]
-
+
```
Comment (Balandat) [inline botorch/models/higher_order_gp.py:292]:
```suggestion

```

**Review by Balandat — COMMENTED** [review]
Let's fix that lint, o/w this lgtm

File: botorch/models/higher_order_gp.py [hunk 3]
```diff
@@ -313,15 +301,16 @@ def rsample(
             ntms_dims = [
                 i == noise_std.shape[0] for i in noiseless_train_marginal_samples.shape
             ]
+            print("ntms_dims", ntms_dims)
```
Comment (Balandat) [inline botorch/models/higher_order_gp.py:304]:
```suggestion
```

File: botorch/models/higher_order_gp.py [hunk 4]
```diff
@@ -313,15 +301,16 @@ def rsample(
...
                     has_matched = True
                 else:
-                    if has_matched:
-                        noise_std = noise_std.unsqueeze(-1)
-                    else:
-                        noise_std = noise_std.unsqueeze(0)
+                    # if has_matched:
+                    noise_std = noise_std.unsqueeze(-1)
+                    # else:
+                    #    noise_std = noise_std.unsqueeze(0)
```
Comment (Balandat) [inline botorch/models/higher_order_gp.py:313]: looks like has_matched isn't doing anything? let's clean this up

**Review by Balandat — APPROVED** [review]
